### PR TITLE
Encapsulate upload classes and provide data verification daemon

### PR DESCRIPTION
**Pull request by ghost**
## Objectives
- Achieve functional and procedural cohesion by encapsulating components involved in the upload process that are responsible for archiving, encryption, upload, and data verification.
- Implement a "Data Verification Daemon" which calls a url after a short delay to confirm validation of uploaded data.
- Provide interfaces for components that enable the API's they wrap (such as OpenSSL) to be easily replaced in their implementation.
- Provide completion blocks which pass errors from the API back to the caller.
- Allow for encryption of data before insertion into an archive.
- Demonstrate use of the encapsulated upload classes using APCDemographicUploader and AsthmaHealth (forthcoming pull request).
- In APCDataArchive, provide methods to schematically derive filenames from ORKFileResult URL's which contain timestamps and unique path extensions.
- The functionality provided by this package duplicates that already provided by APCDataArchiver, APCDataArchiverAndUploader, and Asthma's APHAirQualityDataModel. Forthcoming pull requests will replace upload code in APHAirQualityDataModel, and refactor APCBaseTaskViewController to use these components where possible. 
### Description

The procedure for uploading a file is as follows:
1. Initialize APCDataArchive with a reference that will be the directory name in the temporary directory.
2. Insert files into the archive
3. Initialize an APCDataArchiveUploader and call encryptAndUploadArchive()
4. APCDataArchiveUploader uses composition of APCDataEncryptor, APCDataVerificationClient, and APCDataUploader to upload the data.
5. APCDataUploader incorporates the APCDataVerificationDaemon to report back on data validation.
#### Data Verification Daemon

Implement a daemon as a global variable on APCAppDelegate that queues data verification URL's returned from Sage, and calls the URL for verification after a short delay. Daemon will output a message to the console after 5 seconds of the upload, confirming the upload and validating the data per the database schema. 
### Tests

Tested using APCDemographicUploader and Asthma Air Quality upload.
### Sequence Diagram

uploadsequencediagram
### Class Diagram

uploadclassdiagram

### CODE REVIEW

File: APCAppCore/APCAppCore/APCAppCore.h [hunk 1]
```diff
@@ -77,7 +77,10 @@ FOUNDATION_EXPORT const unsigned char APCAppCoreVersionString[];
  --------------------------------------- */
 #import <APCAppCore/APCDataArchiverAndUploader.h>
 #import <APCAppCore/APCDataArchiver.h>
-#import <APCAppCore/APCPassiveDataCollector.h>
```
Comment (ghost) [inline APCAppCore/APCAppCore/APCAppCore.h:80]: This was a duplicate declaration

File: APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchive.m [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+static NSString * kPhoneInfoKey                     = @"phoneInfo";
+static NSString * kItemKey                          = @"item";
+
+@interface APCDataArchive ()
+
+@property (nonatomic, strong) NSString *reference;
+@property (nonatomic, strong) ZZArchive * zipArchive;
+@property (nonatomic, strong) NSMutableArray * zipEntries;
+@property (nonatomic, strong) NSMutableArray * filesList;
+@property (nonatomic, strong) NSMutableDictionary * infoDict;
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchive.m]: Space between `*` and `variableName` should be consistent.

File: APCAppCore/APCAppCore/UI/TabBarControllers/Profile/APCDemographicUploader.m [hunk 3]
```diff
@@ -109,10 +108,21 @@ - (void)uploadNonIdentifiableDemographicData
         demographics[kPatientWeightPoundsKey] = @(weightInPounds);
     }
     
-    [APCDataArchiverAndUploader uploadDictionary:demographics
-                                withTaskIdentifier:kTaskIdentifierKey
-                                andTaskRunUuid:nil];
-
+    //Archive and upload
+    dispatch_async(dispatch_get_global_queue(QOS_CLASS_BACKGROUND, 0), ^{
+        APCDataArchive *archive = [[APCDataArchive alloc]initWithReference:kTaskIdentifierKey];
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/UI/TabBarControllers/Profile/APCDemographicUploader.m]: Need a space between `]` and `initWithReference`

File: APCAppCore/APCAppCore/Startup/APCAppDelegate.m [hunk 4]
```diff
@@ -174,6 +174,7 @@ - (void)performMigrationAfterDataSubstrateFrom:(NSInteger) __unused previousVers
 
 - (BOOL)application:(UIApplication *) __unused application didFinishLaunchingWithOptions:(NSDictionary *) __unused launchOptions
 {
+    self.dataUploader = [[APCDataUploader alloc]init];
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Startup/APCAppDelegate.m]: Need a space between `]` and `init`

File: APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataVerificationDaemon.m [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                                                userInfo:nil
+                                                 repeats:YES];
+    }
+    
+    return self;
+}
+
+-(void)queueURL:(NSURL *)url
+{
+    APCTimestampedURL *timestampedURL = [[APCTimestampedURL alloc]initWithURL:url];
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataVerificationDaemon.m]: Need a space between `]` and `init`

File: APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataVerificationDaemon.m [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#import "APCDataVerificationDaemon.h"
+#import <BridgeSDK/BridgeSDK.h>
+#import "APCDataVerificationClient.h"
+#import "APCLog.h"
+
+static NSUInteger callDelay = 5;
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataVerificationDaemon.m]: Is this a constant value?

File: APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataUploader.m [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        }
+    }];
+}
+
+#pragma mark - SBBUploadManager delegate
+
+//required delegate method
+- (void)uploadManager:(SBBUploadManager *)__unused manager uploadOfFile:(NSString *)__unused file completedWithError:(NSError *)__unused error
+{
+}
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataUploader.m:87]: At least log the error?

File: APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataUploader.m [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+@implementation APCDataUploader
+
+- (id)init
+{
+    self = [super init];
+    if (self) {
+        SBBUploadManager *uploadManager = [SBBUploadManager defaultComponent];
+        uploadManager.uploadDelegate = self;
+        self.verificationDaemon = [[APCDataVerificationDaemon alloc]init];
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataUploader.m]: Need a space between `]` and `init`

File: APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchive.m [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+
+}
+
++ (NSString *)filenameFromURL: (NSURL *)url
+{
+    NSString *filename = [url lastPathComponent];
+    NSString *pathExtension = [url pathExtension];
+    
+    //Check for known ResearchKit defined filenames with internal path extension
+    NSArray *RKExtensions = [NSArray arrayWithObjects:@"outbound", @"rest", @"walk", nil];
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchive.m]: Prefer modern syntax: `NSArray *RKExtensions = @[@"outbound", @"rest", @"walk"];`

File: APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchive.m [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+                                             kFileInfoTimeStampKey: [NSDate date].toStringInISO8601Format,
+                                             kFileInfoContentTypeKey:[self contentTypeForFileExtension:extension] };
+            
+            [self.filesList addObject:fileInfoEntry];
+            
+        }else {
+            APCLogDebug(@"failed to insert %@ into archive", filename);
+        }
+    }
+}
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchive.m:128]: Consolidate `insertDataAtURLIntoArchive:` and `insertJSONDataIntoArchive:`, I see many common code.

File: APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchive.m [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+    if (![[NSFileManager defaultManager] fileExistsAtPath:workingDirectoryPath]) {
+        NSError * fileError;
+        BOOL created = [[NSFileManager defaultManager] createDirectoryAtPath:workingDirectoryPath withIntermediateDirectories:YES attributes:@{ NSFileProtectionKey : NSFileProtectionComplete } error:&fileError];
+        if (!created) {
+            APCLogError2 (fileError);
+        }
+    }
+    
+    return workingDirectoryPath;
+}
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchive.m:115]: Still return the path even a creation has been failed?

Comment (ghost) [inline APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchive.m:115]: Fixed but not showing here - only in full changes. Set workingDirectoryPath to nil if !created.

File: APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchiveUploader.m [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+-(void)encryptAndUploadArchive:(APCDataArchive *)archive withCompletion:(void (^)(NSError * error))completion
+{
+    __block NSError *error;
+    [archive completeArchiveWithErrorHandler:^(NSError *archiverError) {
+        error = archiverError;
+        
+        if (! archiverError) {
+            //encrypt the archive
+            APCDataEncryptor *encryptor = [[APCDataEncryptor alloc]init];
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchiveUploader.m]: Space between `]` and `init`

File: APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataEncryptor.m [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+{
+    NSError * encryptionError = nil;
+    NSData * unencryptedZipData = [NSData dataWithContentsOfFile:url.relativePath];
+    NSData * encryptedZipData = cmsEncrypt(unencryptedZipData, [APCDataEncryptor pemPath], &encryptionError);
+    
+    if (encryptedZipData) {
+        NSString *encryptedPath = [[self workingDirectoryPath] stringByAppendingPathComponent:kEncryptedDataFilename];
+        
+        if ([encryptedZipData writeToFile:encryptedPath options:NSDataWritingAtomic error:&encryptionError]) {
+            url = [[NSURL alloc]initFileURLWithPath:encryptedPath];
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataEncryptor.m]: Space between ] and init

File: APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataVerificationDaemon.m [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#import "APCDataVerificationDaemon.h"
+#import <BridgeSDK/BridgeSDK.h>
+#import "APCDataVerificationClient.h"
+#import "APCLog.h"
+
+static const NSUInteger callDelay = 5;
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataVerificationDaemon.m]: Shouldn't constant name start with `k`?

**Comment by YuanZhu-apple** [comment]
r=me, merged.

**Pull request merged**